### PR TITLE
Adding BtLE Advertisements

### LINKED ISSUE #85: How to use bluez to act as bluetooth peripheral
The example seems mainly for central bluetooth device. I wonder whether it's possible to use blueZ plugin to implement or to say , emulate a bluetooth peripheral on linux?

### PR DESCRIPTION
**Pull request by RossComputerGuy**
Fixes #85 & #40

This PR adds support for advertisements.

### CODE REVIEW

**Comment by RossComputerGuy** [comment]
```
Unhandled exception:
org.bluez.Error.Failed: Failed to parse advertisement.
#0      _BlueZObject.callMethod (package:bluez/src/bluez_client.dart:1202:11)
<asynchronous suspension>
#1      BlueZAdvertisingManager.registerAdvertisement (package:bluez/src/bluez_client.dart:359:5)
<asynchronous suspension>
#2      main (file:///home/ross/MidstallSoftware/contrib/bluez.dart/example/advertisement.dart:34:3)
<asynchronous suspension>
```
Dang, I know I am close.

**Comment by RossComputerGuy** [comment]
Okay, why is this CLA check not passing? I've signed it half a dozen times.

**Comment by RossComputerGuy** [comment]
Ok, this appears to be working now.

**Comment by RossComputerGuy** [comment]
Added the `unregisterAdvertisement` function. Fully ready and working.

**Comment by robert-ancell** [comment]
> Okay, why is this CLA check not passing? I've signed it half a dozen times.

There'll be a delay with the system updating, thanks!

File: lib/src/bluez_client.dart [hunk 1]
```diff
@@ -167,9 +167,13 @@ final _bluezAddressTypeMap = <String, BlueZAddressType>{
 class BlueZAdapter {
   final String _adapterInterfaceName = 'org.bluez.Adapter1';
 
+  final BlueZClient _client;
   final _BlueZObject _object;
 
-  BlueZAdapter(this._object);
+  BlueZAdapter(this._client, this._object);
+
+  BlueZAdvertisingManager get advertisingManager =>
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:175]: Missing documentation comment.

File: lib/src/bluez_client.dart [hunk 2]
```diff
@@ -340,6 +344,239 @@ class BlueZAdapter {
...
+/// BtLE Advertising Manager
+class BlueZAdvertisingManager {
+  final String _advertInterfaceName = 'org.bluez.LEAdvertisingManager1';
+
+  final BlueZClient _client;
+  final _BlueZObject _object;
+
+  BlueZAdvertisingManager(this._client, this._object);
+
+  Future<void> registerAdvertisement(BlueZAdvertisement advert) async {
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:356]: Missing documentation comment.

File: lib/src/bluez_client.dart [hunk 3]
```diff
@@ -340,6 +344,239 @@ class BlueZAdapter {
...
+    await _client._bus.registerObject(advert);
+
+    await _object.callMethod(_advertInterfaceName, 'RegisterAdvertisement',
+        [advert.path, DBusDict.stringVariant({})],
+        replySignature: DBusSignature(''));
+  }
+
+  Future<void> unregisterAdvertisement(BlueZAdvertisement advert) async {
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:364]: Missing documentation comment.

File: lib/src/bluez_client.dart [hunk 4]
```diff
@@ -340,6 +344,239 @@ class BlueZAdapter {
...
+  }
+}
+
+/// BtLE Advertisement instance
+abstract class BlueZAdvertisement extends DBusObject {
+  final String _advertInterfaceName = 'org.bluez.LEAdvertisement1';
+
+  BlueZAdvertisement(DBusObjectPath path) : super(path);
+
+  Future<void> release();
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:379]: Missing documentation comment.

File: lib/src/bluez_client.dart [hunk 5]
```diff
@@ -340,6 +344,239 @@ class BlueZAdapter {
...
+
+  Map<BlueZManufacturerId, DBusValue> get manufacturerData;
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:381]: I think these should be constructor parameters, i.e. `BlueZAdvertisement(this.type, this.serviceUuids, ...)`. These values never change so they don't need to be accessed via getters.

File: lib/src/bluez_client.dart [hunk 6]
```diff
@@ -1292,7 +1529,7 @@ class BlueZClient {
 
   /// Creates a new BlueZ client. If [bus] is provided connect to the given D-Bus server.
   BlueZClient({DBusClient? bus})
-      : _bus = bus ?? DBusClient.system(),
+      : _bus = bus ?? DBusClient.system(introspectable: true),
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:1532]: Can you put this in a separate PR.

File: lib/src/bluez_client.dart [hunk 7]
```diff
@@ -340,6 +344,239 @@ class BlueZAdapter {
...
+/// BtLE Advertisement instance
+abstract class BlueZAdvertisement extends DBusObject {
+  final String _advertInterfaceName = 'org.bluez.LEAdvertisement1';
+
+  BlueZAdvertisement(DBusObjectPath path) : super(path);
+
+  Future<void> release();
+
+  Map<BlueZManufacturerId, DBusValue> get manufacturerData;
+  String get type;
```
Comment (robert-ancell) [inline lib/src/bluez_client.dart:382]: Are the types fixed? If so change to an enum.

**Review by robert-ancell — CHANGES_REQUESTED** [review]
Thanks, I think with the requested changes this should be good.

File: lib/src/bluez_client.dart (hunk 6 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:1532]: Leftover from some skill issue with trying to get this PR to work, unnecessary change and I'll remove it.

File: lib/src/bluez_client.dart (hunk 7 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:382]: Yes, I'll make it an enum.

Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: Done

File: lib/src/bluez_client.dart (hunk 6 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: Removed.

File: lib/src/bluez_client.dart (hunk 5 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: I added a lot to the constructor, any others which should be added?

File: lib/src/bluez_client.dart (hunk 4 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: Done

File: lib/src/bluez_client.dart (hunk 3 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:390]: Done

File: lib/src/bluez_client.dart (hunk 2 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:377]: Done

File: lib/src/bluez_client.dart (hunk 1 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart:176]: Done

File: lib/src/bluez_client.dart (hunk 5 above)
Comment (RossComputerGuy) [reply lib/src/bluez_client.dart]: Okay, everything is a constructor.

File: example/advertisement.dart [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+void main() async {
+  var client = BlueZClient();
+  await client.connect();
+
+  var adapter = client.adapters.first;
+  var advertMngr = adapter.advertisingManager;
+
+  var advert = BlueZAdvertisement(
+    DBusObjectPath('/org/bluez/example/advertisement0099'),
```
Comment (robert-ancell) [inline example/advertisement.dart:12]: We don't want to provide this.

File: example/advertisement.dart [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        0x01,
+        0x00,
+        0x02,
+        0x0c
+      ]),
+    },
+    localName: 'Example Advert',
+  );
+
+  await advertMngr.registerAdvertisement(advert);
```
Comment (robert-ancell) [inline example/advertisement.dart:44]: This should be:
```dart
var advert = await advertMngr.registerAdvertisement(type: BlueZAdvertisementType.broadcast,
    manufacturerData: {
      BlueZManufacturerId(0x004c): DBusArray.byte([
        0x02,
        0x15,
        0xE2,
        0xC5,
        0x6D,
        0xB5,
        0xDF,
        0xFB,
        0x48,
        0xD2,
        0xB0,
        0x60,
        0xD0,
        0xF5,
        0xA7,
        0x10,
        0x96,
        0xE0,
        0x00,
        0x01,
        0x00,
        0x02,
        0x0c
      ]),
    },
    localName: 'Example Advert');
```

In this way the path is allocated by the manager and can't collide.

**Review by robert-ancell — CHANGES_REQUESTED** [review]

File: example/advertisement.dart (hunk 8 above)
Comment (RossComputerGuy) [reply example/advertisement.dart]: Alright, it is automatically generated now.

File: example/advertisement.dart (hunk 9 above)
Comment (RossComputerGuy) [reply example/advertisement.dart]: Done

**Review by robert-ancell — APPROVED** [review]
Awesome work, thanks!
